fix: Composer nav + button discoverability

lei9444 opened the pull request:
## Description
new Design: https://www.figma.com/file/deFhMPBnRWxTYiYDYqYxrc/Composer---Vivek?node-id=10809%3A0
![image](https://user-images.githubusercontent.com/39758135/107217408-173d9200-6a49-11eb-890e-6633c6540f5d.png)

![filter1](https://user-images.githubusercontent.com/39758135/107206595-34b72f80-6a3a-11eb-8386-6ce3956fad2c.gif)

This PR also fix one bug. The bug is the filter is used in some function, and the function can only get the stale value, so when we input the text into the search box, The filter doesn't work
![image](https://user-images.githubusercontent.com/39758135/107389170-855d8400-6b31-11eb-9157-304332ce4aa8.png)
In this PR use Ref to make sure the value is latest.

<!---
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.

If this is a bug fix, please describe the root cause and analysis of this problem.
---->

## Task Item
closes #5680 
<!---
Please include a link to the related issue. [Ex. `Closes #<issue #>`](https://help.github.com/en/articles/closing-issues-using-keywords)
---->

## Screenshots

<!---
Please include screenshots or gifs if your PR include UX changes.
--->

hatpick commented:
Does this PR also solve the following issue:
(header of project tree is scrolling with it)
https://user-images.githubusercontent.com/1483492/107267236-74aaf200-69fb-11eb-98fe-25a59ecddde4.mov

liweitian commented:
> Does this PR also solve the following issue:
> (header of project tree is scrolling with it)
> https://user-images.githubusercontent.com/1483492/107267236-74aaf200-69fb-11eb-98fe-25a59ecddde4.mov

From the design, the header is removed. @vivekkshankar could you help verify on this? The search bar is also changed on lu/lg/qna page. Should we do that?

hatpick commented on Composer/packages/client/src/components/ProjectTree/ProjectTreeHeader.tsx:
```diff
@@ -51,54 +118,47 @@ export const ProjectTreeHeader: React.FC<ProjectTreeHeaderProps> = ({ menu }) =>
+          onChange={onFilter}
+        />
+      ) : (
         <OverflowSet
```
why do you need this overflow set? Simply used a commandbar button
(https://developer.microsoft.com/en-us/fluentui#/controls/web/button)

hatpick commented on Composer/packages/client/src/components/ProjectTree/ProjectTree.tsx:
```diff
@@ -197,6 +187,8 @@ export const ProjectTree: React.FC<Props> = ({
   const setPageElement = (name: string, value) => setPageElementState('dialogs', { ...pageElements, [name]: value });
 
   const [filter, setFilter] = useState('');
+  const filterRef = useRef(filter);
```
I dont understand what's happening with this ref, what does it represent? and why is it needed?

lei9444 replied:
Hi, @hatpick , There is bug in main branch, I fixed it in this PR. The bug is the filter doesn't work when we input some text. The reason is I found the filter text is used in a function, but react will use the stale value anytime
![image](https://user-images.githubusercontent.com/39758135/107379522-f730d000-6b27-11eb-8c21-b498c0ce077b.png)

hatpick replied:
I have created a hook in my local branch for another feature:
```
export const useDebounce = <T>(value: T, delay: number) => {
  const [debouncedValue, setDebouncedValue] = useState<T>(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [value]);

  return debouncedValue;
};
```
you can consume it for this exact case like:
```
const [query, setQuery] = React.useState<string | undefined>();
const debouncedQuery = useDebounce<string | undefined>(query, 300);
```

and then you deal with ```debouncedQuery```

wanna give it a try? if it works as expected you can add that hook to share library so others can use it too.

lei9444 replied:
Thanks @hatpick , I think I didn't explain the issue very clearly. 
I think there is a wrong usage in < ProjectTree />: Some children components will use the filter message to re-render the UI, but all these components will call a function from parent to filter their items and they don't depends on the filter message change, so the function will only use the stale value. 

So I think I just need to listen to the filter state and re-render these components when filter change. I fixed it in the latest commit and removed the Ref.

I think it's a good idea to share the useDebounce to the library, the filter state in ProjectTree also use a  similar implementation with throttle. Can we do this in a new PR?

hatpick commented on Composer/packages/client/src/components/ProjectTree/ProjectTreeHeader.tsx:
```diff
@@ -34,9 +51,59 @@ export interface ProjectTreeHeaderMenuItem {
 
 export interface ProjectTreeHeaderProps {
   menu: ProjectTreeHeaderMenuItem[];
+  onFilter?: (_e, newValue?: string) => void;
```
if you dont want to use the event why are you including it in the callback?

hatpick commented on Composer/packages/client/src/components/ProjectTree/ProjectTreeHeader.tsx:
```diff
@@ -3,27 +3,44 @@
+`;
 
 const headerText = css`
+  border-bottom: 1px solid #edebe9;
```
use a color from Fluent instead of hard coding it

hatpick commented on Composer/packages/client/src/components/ProjectTree/ProjectTreeHeader.tsx:
```diff
@@ -3,27 +3,44 @@
-import { moreButton, overflowSet } from './treeItem';
+const searchBox: ISearchBoxStyles = {
+  root: {
+    borderBottom: '1px solid #edebe9',
```
same here for the color
https://developer.microsoft.com/en-us/fluentui#/styles/web/colors/neutrals
gray30

hatpick commented on Composer/packages/client/src/components/ProjectTree/ProjectTreeHeader.tsx:
```diff
@@ -34,9 +51,59 @@ export interface ProjectTreeHeaderMenuItem {
+const onRenderOverflowButton = (isActive: boolean) => {
+  const moreLabel = formatMessage('Actions');
+  return (overflowItems: IContextualMenuItem[] | undefined) => {
+    if (overflowItems == null) return null;
```
if(!overflowItems) {
   return null;
}

nit: avoid using Abstract Equality Comparison (==) vs. ===

hatpick commented:
> > Does this PR also solve the following issue:
> > (header of project tree is scrolling with it)
> > https://user-images.githubusercontent.com/1483492/107267236-74aaf200-69fb-11eb-98fe-25a59ecddde4.mov
> 
> @hatpick From the design, the header is removed. @vivekkshankar could you help verify on this? The search bar is also changed on lu/lg/qna page. Should we do that?

When the project tree list is long and user scrolls, the "header" that you have add button and filtering in it will scroll oit of view with it, @vivekkshankar is that what you wanted for ux?

lei9444 commented:
> > > Does this PR also solve the following issue:
> > > (header of project tree is scrolling with it)
> > > https://user-images.githubusercontent.com/1483492/107267236-74aaf200-69fb-11eb-98fe-25a59ecddde4.mov
> > 
> > 
> > @hatpick From the design, the header is removed. @vivekkshankar could you help verify on this? The search bar is also changed on lu/lg/qna page. Should we do that?
> 
> When the project tree list is long and user scrolls, the "header" that you have add button and filtering in it will scroll out of the viewport, @lei9444 @vivekkshankar is that what you wanted for ux?

Hi, @hatpick I think we need to fix the header, I have fixed in this PR 
![image](https://user-images.githubusercontent.com/39758135/107385398-d23f5b80-6b2d-11eb-9d27-8f98a72e1e85.png)

srinaath commented:
@hatpick Are we good with this PR? It removes a P0 of our plate.
